src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -409,11 +401,118 @@ class RegionValueBuilder(region: MemoryBuffer) {
 
   def addRow(t: TStruct, r: Row) {
     assert(r != null)
+
+    startStruct()
     var i = 0
     while (i < t.size) {
       addAnnotation(t.fields(i).typ, r.get(i))
       i += 1
     }
+    endStruct()
+  }
+
+  def fixupBinary(toOff: Int, fromRegion: MemoryBuffer, fromOff: Int) {
+    val fromBOff = fromRegion.loadInt(fromOff)
+    val length = fromRegion.loadInt(fromBOff)
+
+    region.align(4)
+    val toBOff = region.allocate(4 + length)
+    region.storeInt(toOff, toBOff)
+
+    region.copyFrom(fromRegion, fromBOff, toBOff, 4 + length)
+  }
+
+  def requiresFixup(t: Type): Boolean = {
```

Comment:
Yes.  See the entrypoint into the fixup routines, like `addUnsafeRow`, which calls `fixupStruct`:

```
    fixupStruct(t.fundamentalType, toOff, fromRegion, fromOff)
```

Hmm, yeah, I think I like that.  I'll make that change.